infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/log/ContainerLogWatch.java

```diff
@@ -0,0 +1,255 @@
+import com.google.gson.JsonParseException;
+import com.google.gson.JsonParser;
+import io.fabric8.kubernetes.client.KubernetesClient;
+import io.fabric8.kubernetes.client.dsl.LogWatch;
+import java.io.BufferedReader;
+import java.io.Closeable;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.util.Objects;
+import java.util.StringJoiner;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This class is responsible for watching logs in a single container instance. Messages are
+ * provided, unbuffered, to a provided {@link PodLogHandler}
+ *
+ * <p>Reading logs is a blocking operation.
+ */
+class ContainerLogWatch implements Runnable, Closeable {
+
+  private static final Logger LOG = LoggerFactory.getLogger(ContainerLogWatch.class);
+
+  private final KubernetesClient client;
+  private final PodLogHandler logHandler;
+  private final LogWatchTimeouts timeouts;
+
+  private final String namespace;
+  private final String podName;
+  private final String containerName;
+
+  // current LogWatch instance. We need it so we can close it from outside in close() method.
+  private LogWatch currentLogWatch;
+
+  // json parser used to parse log messages to check for errorness
+  private final JsonParser jsonParser = new JsonParser();
+
+  // flag whether we should still try to get the logs
+  private final AtomicBoolean closed = new AtomicBoolean(false);
+
+  ContainerLogWatch(
+      KubernetesClient client,
+      String namespace,
+      String podName,
+      String containerName,
+      PodLogHandler logHandler,
+      LogWatchTimeouts timeouts) {
+    this.client = client;
+    this.namespace = namespace;
+    this.podName = podName;
+    this.containerName = containerName;
+    this.logHandler = logHandler;
+    this.timeouts = timeouts;
+  }
+
+  /**
+   * Do the best effort to get the logs from the container. The method is trying for {@link
+   * LogWatchTimeouts#getWatchTimeoutMs()} to get the logs from the container. If response on log
+   * request from the k8s is 40x, it possibly means that container is not ready to get the logs and
+   * we'll try again after {@link LogWatchTimeouts#getWaitBeforeNextTry()} milliseconds.
+   */
+  @Override
+  public void run() {
+    Stopwatch stopwatch = Stopwatch.createStarted();
+    // we try to get logs for `timeouts.getWatchTimeoutMs()`
+    while (stopwatch.elapsed(TimeUnit.MILLISECONDS) < timeouts.getWatchTimeoutMs()) {
+
+      // request k8s to get the logs from the container
+      try (LogWatch logWatch =
+          client
+              .pods()
+              .inNamespace(namespace)
+              .withName(podName)
+              .inContainer(containerName)
+              .watchLog()) {
+
+        // we need to synchroinze here to avoid adding new `currentLogWatch` after we close it
+        synchronized (this) {
```

Comment:
unexpected to me to see the usage of synchronized + AtimicBoolean. What can bad happen if we remove it?

Reply:
yes, we don't need AtomicBoolean here. replaced with regular boolean